libdrgn/iterators.h

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#ifndef DRGN_ITERATORS_H
+#define DRGN_ITERATORS_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include "drgn.h"
+#include "vector.h"
+
+struct radix_tree_iter_frame {
+	struct drgn_object slots;
+	uint64_t index;
+	uint64_t shift;
+	uint64_t loop_index;
+};
+
+DEFINE_VECTOR_TYPE(frames, struct radix_tree_iter_frame)
+
+typedef struct radix_tree_iter_entry {
+	uint64_t index;
+	struct drgn_object node;
+} radix_tree_iter_entry_t;
+
+struct radix_tree_iter {
+	bool started;
+	struct drgn_object root;
+	// Current value to be yielded
+	struct radix_tree_iter_entry entry;
+	// We need this for later initialization of `drgn_object`s
+	struct drgn_program *prog;
+	// Frames to keep track of generator state
+	struct frames frames;
+	// One-time setup values that are persistent
+	uint64_t RADIX_TREE_INTERNAL_NODE;
+	uint64_t RADIX_TREE_MAP_MASK;
+	struct drgn_qualified_type node_type;
+};
+
+struct drgn_error *radix_tree_iter_init(struct drgn_program *prog,
+					const struct drgn_object *root,
+					struct radix_tree_iter *ret);
+
+struct drgn_error *radix_tree_iter_next(struct radix_tree_iter *iter,
+					struct radix_tree_iter_entry **ret);
+
+void radix_tree_iter_deinit(struct radix_tree_iter *iter);
+
+typedef struct radix_tree_iter_entry idr_iter_entry_t;
+
+struct idr_iter {
+	struct radix_tree_iter iter;
+	uint64_t base;
+};
+
+struct drgn_error *idr_iter_init(const struct drgn_object *idr,
+				 struct idr_iter *ret);
+
+struct drgn_error *idr_iter_next(struct idr_iter *iter, idr_iter_entry_t **ret);
+
+void idr_iter_deinit(struct idr_iter *iter);
+
+typedef struct drgn_object pid_iter_entry_t;
+
+struct pid_iter {
+	bool has_idr;
+	struct drgn_qualified_type pid_type;
+	union {
+		// if has_idr
+		struct idr_iter iter;
+		// else
+		struct {
+			struct drgn_qualified_type upid_type;
+			struct drgn_object pid_hash;
+			struct drgn_object pos; // a `struct hlist_node*`
+			struct drgn_object ns;
+			struct drgn_object entry; // Current value of the iterator
+			int index; // Current loop index
+			size_t hash_length; // 1 << pidhash_shift
+			char member_specifier[9 + 19 + 1];
+			// 9 = length of "numbers[]"
+			// 19 = maximum length of a uint64_t as a string
+			// 1 = space for the null terminator
```

Comment:
You can make this a little more self-documenting by using `char member_specifier[sizeof("numbers[]") + 19]`. Note that `sizeof` includes the null terminator. Also, should this be 20?

```
$ python3 -c 'print(len(str(2**64 - 1)))'
20
```

There are also clever tricks you can use to get the maximum number of decimal digits for an arbitrary integer type: https://stackoverflow.com/questions/10536207/ansi-c-maximum-number-of-characters-printing-a-decimal-int. But that's probably overkill.